streaming/src/main/java/org/apache/spark/streaming/util/WriteAheadLog.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.streaming.util;
+
+import java.nio.ByteBuffer;
+import java.util.Iterator;
+
+/**
+ * Interface representing a write ahead log (aka journal) that is used by Spark Streaming to
+ * save the received data (by receivers) and associated metadata to a reliable storage, so that
+ * they can be recovered after driver failures. See the Spark docs for more information on how
+ * to plug in your own custom implementation of a write ahead log.
+ */
+@org.apache.spark.annotation.DeveloperApi
+public interface WriteAheadLog {
```

Comment:
Yes. Its meant for users to create arbitrary implementations and we want to
stay backward compatible (scala traits  have pretty nasty corner cases).
On Apr 26, 2015 9:48 PM, "Hari Shreedharan" <notifications@github.com>
wrote:

> In
> streaming/src/main/java/org/apache/spark/streaming/util/WriteAheadLog.java
> <https://github.com/apache/spark/pull/5645#discussion_r29120494>:
>
> > + * limitations under the License.
> > + */
> > +
> > +package org.apache.spark.streaming.util;
> > +
> > +import java.nio.ByteBuffer;
> > +import java.util.Iterator;
> > +
> > +/**
> > + * Interface representing a write ahead log (aka journal) that is used by Spark Streaming to
> > + * save the received data (by receivers) and associated metadata to a reliable storage, so that
> > + * they can be recovered after driver failures. See the Spark docs for more information on how
> > + * to plug in your own custom implementation of a write ahead log.
> > + */
> > +@org.apache.spark.annotation.DeveloperApi
> > +public interface WriteAheadLog {
>
> Is the idea that this would be useful for Java implementations to keep
> this a Java interface?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/5645/files#r29120494>.
>